npm/src/worker-proxy.ts

```diff
@@ -0,0 +1,387 @@
+): TService &
+  IServiceProxy & {
+    onMsgFromWorker: (
+      msg: ResponseMessage<TService> | EventMessage<TServiceEventMsg>
+    ) => void;
+  } {
+  const queue: RequestState<TService, TServiceEventMsg>[] = [];
+  const eventTarget = new EventTarget();
+  let curr: RequestState<TService, TServiceEventMsg> | undefined;
+  let state: ServiceState = "idle";
+
+  function setState(newState: ServiceState) {
+    if (state === newState) return;
+    state = newState;
+    if (proxy.onstatechange) proxy.onstatechange(state);
+  }
+
+  type ResultOf<TRespMsg> = TRespMsg extends { result: infer R } ? R : never;
+
+  function queueRequest(
+    msg: RequestMessage<TService>,
+    requestEventTarget?: IServiceEventTarget<TServiceEventMsg>,
+    cancellationToken?: CancellationToken
+  ): Promise<ResultOf<ResponseMessage<TService>>> {
+    return new Promise((resolve, reject) => {
+      queue.push({
+        type: msg.type,
+        args: msg.args,
+        resolve,
+        reject,
+        requestEventTarget,
+        cancellationToken,
+      } as RequestState<TService, TServiceEventMsg>);
+
+      // If nothing was running when this got added, kick off processing
+      if (queue.length === 1) doNextRequest();
+    });
+  }
+
+  function doNextRequest() {
+    if (curr) return;
+
+    while ((curr = queue.shift())) {
+      // eslint-disable-line no-cond-assign
+      if (curr.cancellationToken?.isCancellationRequested) {
+        curr.reject("cancelled");
+        continue;
+      } else {
+        break;
+      }
+    }
+    if (!curr) {
+      // Nothing else queued, signal that we're now idle and exit.
+      log.debug("Proxy: Worker queue is empty");
+      setState("idle");
+      return;
+    }
+
+    const msg = { type: curr.type, args: curr.args };
+    if (methods[curr.type] === "requestWithProgress") {
+      setState("busy");
+    }
+
+    log.debug("Proxy: Posting message to worker: %o", msg);
+    postMessage(msg);
+  }
+
+  function onMsgFromWorker(
+    msg: ResponseMessage<TService> | EventMessage<TServiceEventMsg>
+  ) {
+    if (log.getLogLevel() >= 4)
+      log.debug("Proxy: Received message from worker: %s", JSON.stringify(msg));
+
+    if (msg.messageType === "event") {
+      const event = new Event(msg.type) as Event & TServiceEventMsg;
+      event.detail = msg.detail;
+
+      log.debug("Proxy: Posting event: %o", msg);
+      // Post to a currently attached event target if there's a "requestWithProgress"
+      // in progress
+      curr?.requestEventTarget?.dispatchEvent(event);
+      // Also post to the general event target
+      eventTarget.dispatchEvent(event);
+    } else if (msg.messageType === "response") {
+      if (!curr) {
+        log.error("Proxy: No active request when message received: %o", msg);
+        return;
+      }
+      const result = {
+        success: msg.result.success,
+        data: msg.result.success ? msg.result.result : msg.result.error,
+      };
+      if (!result) return;
```

Comment:
Was this line here before? Seems a little odd. Based on the prior statement, how would this condition ever be true?

Reply:
Oops, you're right, it's a leftover from refactoring. I originally had the code above it extracted to a function that may return `undefined`. I'll remove this line.